specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+* WNS will ensure that the payload is ultimately delivered to the app residing on the client device.
+
+Phase 1 (Reunion Preview):
+WNS Push APIs will only support WIN32 Packaged app (MSIX).
+
+Phase 2 (Reunion V1):
+Our Push APIs will add support for Unpackaged Win32 scenarios.
+
+# Examples
+
+## Registration of the Push Activator for Win32 (Inproc)
+If a Win32 app needs to subscribe to Push triggers inproc, the code in Main would follow the pattern below:
+* The Registration will take in a CLSID of the COM server as part of the Activator setup. The registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and abstract away the COM details from the developer.
+* The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a seperate AppLifeCycle API spec.)
+* If the Activation is Push, the app will QI the Arguments property to retrieve an instance of PushActivatedEventArgs and get the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+  * It will subscribe to a Foreground In-memory Push event
+  * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler.
+    * Expect Channel operations to take around 2-16 minutes in some rare cases (retryable errors). In the 90th percentile case, it should be fairly quick operation (in a few seconds).  
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    winrt::guid activatorGuid{ 0xC380465D, 0x2271, 0x428C, { 0x9B, 0x83, 0xEC, 0xEA, 0x3B, 0x4A, 0x85, 0xC1} };
+    InProcActivatorDetails details(activatorGuid);
+    PushManager::RegisterActivator(details);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    // Note: This API is currently not present in Reunion but will be included as part of the AppLifecycle Public API spec.
+    auto activation = AppLifecycle::Activation();
+    auto kind = activation.Kind();
+
+    if (kind == ReunionActivationKind::Push)
+    {
+        auto pushArgs = activation.Arguments().try_as<PushActivatedEventArgs>();
+
+        // Call TakeDeferral to ensure that code runs in low power
+        pushArgs.GetDeferral();
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ReunionActivationKind::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushActivated([](const auto&, PushActivatedEventArgs args)
```

Comment:
`PushActivated` sounds a bit ambiguous, more like the event when your app is activated because of a push. In our current SDK, we use the terminology "Received" for this foreground event handler, so maybe `PushReceived` would be more clear? Idk, it's probably fine either way, but something about "Activated" feels stranger to me!

Reply:
I'm good with PushReceived. Will make that change.

Reply:
@andrewleader API review feedback has been to append everything with PushNotification*. For example PushNotificationChannel, PushNotificationReceived, PushNotificationManager. So this would need to be tweaked based on the new suffix.

Reply:
Only type names need to have the `PushNotification` prefix. Members, events, fields do not need the prefix since they are already scoped to the containing type and will not create name conflicts.

So the event can still be called `PushReceived` because it's part of `PushActivatorDetails`. You don't have to worry about conflicting with other members of `PushActivatorDetails` since you control all of them.

The prefix naming is used on types, because your `Channel` runtime class could conflict with a `Channel` class in the VPN namespace. It is common to use `using` statements to erase namespace boundaries.

Reply:
@oldnewthing
Is the naming convention for properties similar to events then?  What about functions? Should it just be CreateChannelAsync or CreatePushChannelAsync or CreatePushNotificationChannelAsync? They are all under the PushNotificationChannelManager type.

Reply:
Names that are scoped to a type need only be unique within the type.

```cs
enum HolographicMuffinColor
{
  Blue, // short name allowed and encouraged
  HolographicMuffinColorBlue, // long name unnecessary and discouraged
}

runtimeclass HolographicMuffin
{
   HolographicMuffinColor Color{ get; }; // short name allowed and encouraged
   HolographicMuffinColor HolographicMuffinSize{ get; }; // long name unnecessary and discouraged

   void SetColor(HolographicMuffinColor color); // short name allowed and encouraged
   void SetHolographicMuffinColor(HolographicMuffinColor color); // long name unnecessary and discouraged

   event TypedEventHandler<HolographicMuffin, Object> ColorChanged; // short name allowed and encouraged
   event TypedEventHandler<HolographicMuffin, Object> HolographicMuffinColorChanged;  // long name unnecessary and discouraged
}

```

Type names must be unique across namespaces because it is common to erase namespace boundaries by liberal use of the `using` statement.

```cs
namespace Windows.Graphics {
    runtimeclass Bitmap; // a generic pixel bitmap
}
namespace Windows.Devices.PointOfService.LineDisplays
{
    runtimeclass Bitmap; // a bitmap on a line display
}

namespace Windows.UI.Xaml.Media
{
    runtimeclas Bitmap; // a XAML bitmap
}

// An app that uses XAML to display images on a Point of Service line display

using Windows.Graphics;
using Windows.Devices.PointOfService.LineDisplays;
using Windows.UI.Xaml.Media;

Bitmap bitmap = new Bitmap(); // which bitmap?
```